text/XXX-unwind-FFI.md

```diff
@@ -0,0 +1,286 @@
+- Feature Name: `unwind-through-FFI`
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- RFC PR: ???
+- Rust Issue: ???
+
+# Summary
+[summary]: #summary
+
+Provide a well-defined mechanism for unwinding through FFI boundaries.
+
+* Add an `#[unwind(allowed)]` function attribute that's supported on `extern` functions and `extern` function declarations.
+  which explicitly permits `extern` functions
+  to unwind (`panic`) without aborting.
+* If this annotation is used anywhere in the dependency tree, 
```

Comment:
@kornelski In that case, I am somewhat confused as to why it's important to have a way to permit `panic`s to propagate past FFI boundaries. In [the post that called my attention to the problem](https://internals.rust-lang.org/t/unwinding-through-ffi-after-rust-1-33/9521?u=batmanaod), you wrote:

> These two popular libraries are specifically designed to be unwound....
>
> Rust 1.33 effectively killed my `mozjpeg` crate...

If aborting is "perfectly fine", why does `mozjpeg` need a way to unwind past FFI boundaries?

If indeed `libjpeg` wrappers do need to let unwind operations cross FFI boundaries, that _is already_ a global configuration issue that must be solved for the whole binary (which may be a dynamic library rather than a final executable). This is because such unwinding _is not safe_ unless the behavior of the unwind runtime is stable and guaranteed, and the default `panic` runtime does not and will not provide such a guarantee.

Reply:
@kornelski I think part of the problem is what happens if you compile the app with `panic=abort` but the dynamic lib was compiled with `panic=unwind`. In that case the unwind machinery won't exist. I can think of two ways to solve this:

1. Add a feature to the (`mozjpeg`) lib so users can opt out of `#[unwind(allowed)]` attributes if they binary is being compiled with `panic=abort`. I'm not sure how this would work if `mozjpeg` is an indirect dependency... I haven't ever tried to fiddle with a transitive denedency's feature flags.
2. Modify this RFC so it supports `#[unwind(allowed)]` with `panic=abort`. This would require the compiler to implicitly wrap calls to the extern panic-able function in a catch block to catch any exception and immediately call abort. I'm not sure whether or not this is feasible in all situations (e.g., weird `no_std` embedded environments).

Reply:
@mjbshaw I don't understand your second suggestion. If a dependency graph has a crate with `panic=abort` in it, then all `panic`s will abort, and, if `#[unwind(allowed)]` is permitted (as it currently is in Nightly), it will have no effect whatsoever.

As for `no_std` environments, those currently can't `panic` without a custom `panic_runtime` crate (which is the main reason `#![panic_runtime]` exists).